opm/simulators/wells/BlackoilWellModel_impl.hpp

```diff
@@ -1229,7 +1234,144 @@ namespace Opm {
+                                                         gratTargetFromSales, nodeName, group_state,
+                                                         group.has_gpmaint_control(cmode));
+                double orig_target = tcalc.groupTarget(ctrl, local_deferredLogger);
+
+                auto mismatch = [&] (auto well_group_thp) {
+                    double well_group_rate(0.0);
+                    double rate(0.0);
+                    for (auto& well : this->well_container_) {
+                        std::string well_name = well->name();
+                        auto& ws = well_state.well(well_name);
+                        if (group.hasWell(well_name)) {
+                            rate = -tcalc.calcModeRateFromRates(ws.surface_rates);
+                            well->setDynamicThpLimit(well_group_thp);
+                            const Well& well_ecl = wells_ecl_[well->indexOfWell()];
+                            const auto inj_controls = Well::InjectionControls(0);
+                            const auto prod_controls = well_ecl.productionControls(summary_state);
+                            well->iterateWellEqWithSwitching(this->ebosSimulator_, dt, inj_controls, prod_controls, well_state, group_state, local_deferredLogger,  false, false);
+                            rate = -tcalc.calcModeRateFromRates(ws.surface_rates);
+                            if (ws.production_cmode == Opm::WellProducerCMode::THP)
+                                ws.thp = well_group_thp;
+                            well_group_rate += rate;
+                        }
+                    }
+                    double diff = well_group_rate - orig_target;
+                    return diff;
+                };
+
+                double well_group_rate(0.0);
+                double rate(0.0);
+                double thp(0.0);
+                double min_thp(1.0E8);
+                double max_thp(0.0);
+                for (auto& well : this->well_container_) {
+                    std::string well_name = well->name();
+                    if (group.hasWell(well_name)) {
+                        auto& ws = well_state.well(well_name);
+                        rate = -tcalc.calcModeRateFromRates(ws.surface_rates);
+                        well_group_rate += rate;
+                        thp = ws.thp;
+                        min_thp = std::min(min_thp, thp);
+                        max_thp = std::max(max_thp, thp);
+                    }
+                }
 
+                const auto upbranch = network.uptree_branch(nodeName);
+                const auto it = node_pressures_.find((*upbranch).uptree_node());
+                const double nodal_pressure = it->second;
+                double well_group_thp = nodal_pressure;
+
+                if ((reportStepIdx <= 1) || (well_group_rate > 0.9*orig_target)) {
+                    std::array<double, 2> range;
+                    if (well_group_thp_prev_ == 0) {
```

Comment:
safe range for the THP is very important, we should discuss regarding the current way of determining the search range for THP.

Reply:
I am struggling with this. @GitPaean It would be good to have a chat about this.